tests/main/snapshot-exclusions-dynamic/task.yaml

```diff
@@ -0,0 +1,132 @@
+summary: Check that the basic dynamic snapshot data exclusion functionality is ok
+
+environment:
+    # tar invoked when creating snapshots triggers OOM
+    SNAPD_NO_MEMORY_LIMIT: 1
+    USER_DATA_DIR: /root/snap/test-snap
+    SYSTEM_DATA_DIR: /var/snap/test-snap
+
+prepare: |
+    "$TESTSTOOLS"/snaps-state install-local test-snap
+
+restore: |
+    rm -f expected_options_entry actual_options_entry expected_restored_files actual_restored_files ./*_export.snapshot
+
+debug: |
+    snap saved || true
+
+execute: |
+    # waitAsyncChangeDone() waits for completion of async change
+    waitAsyncChangeDone() {
+        local change isDone
+        change=$1
+        isDone=0
+        while [ "$isDone" -ne 1 ]
+        do
+            isDone=$( curl -s --unix-socket /run/snapd.socket http://localhost/v2/changes/"$change" | jq '.result.tasks[].progress.done' )
+            sleep 0.01
+        done
+    }
+
+    # Use the snap, so that the data dirs will be created
+    test-snap.sh -c 'true'
+
+    # Drop in canary files that should never be excluded
+    echo "system versioned" > "$SYSTEM_DATA_DIR/current/canary.txt"
+    echo "system common" > "$SYSTEM_DATA_DIR/common/canary.txt"
+    echo "user versioned"  > "$USER_DATA_DIR/current/canary.txt"
+    echo "user common" > "$USER_DATA_DIR/common/canary.txt"
+    
+    # Create files to be statically and dynamically excluded on all four data-locations
+    for DIR in "$USER_DATA_DIR"/{current,common} "$SYSTEM_DATA_DIR"/{current,common}
+    do
+        touch "$DIR/static-exclude.txt"
+        touch "$DIR/dynamic-exclude.txt"
+    done
+
+    # Create snapshot that will apply static exclusions, and grab the set ID
+    SET_ID=$( snap save test-snap | cut -d\  -f1 | tail -n1 )
+    echo "$SET_ID" | MATCH "^[0-9]+$"
+    
+    # Confirm the snapshot for the set ID includes the test-snap
+    snap saved --id="$SET_ID" | MATCH "test-snap"
+    # and is valid
+    snap check-snapshot "$SET_ID"
+    # and the snapshot metadata does not include an options entry
+    curl -s --unix-socket /run/snapd.socket --max-time 5 http://localhost/v2/snapshots?set="$SET_ID" | NOMATCH '"options":'
+
+    # Create snapshot that will apply dynamic exclusions, and grab the set ID
+    RESPONSE=$( sudo curl -s --unix-socket /run/snapd.socket --max-time 5 http://localhost/v2/snaps -X POST -H 'Content-Type: application/json' -d '{"action": "snapshot", "snaps": ["test-snap"], "snapshot-options": {"test-snap":{"exclude":["$SNAP_DATA/dynamic-exclude.txt", "$SNAP_COMMON/dynamic-exclude.txt", "$SNAP_USER_COMMON/dynamic-exclude.txt", "$SNAP_USER_DATA/dynamic-exclude.txt"]}}}' )
```

Comment:
sudo is not needed (spread runs as root)

Reply:
oops, copy/paste... removed